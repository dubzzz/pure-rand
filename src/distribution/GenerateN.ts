import type { RandomGenerator } from '../types/RandomGenerator';
import { unsafeGenerateN } from './UnsafeGenerateN';

export function generateN(rng: RandomGenerator, num: number): [number[], RandomGenerator] {
  const nextRng = rng.clone();
  const out = unsafeGenerateN(nextRng, num);
  return [out, nextRng];
}
