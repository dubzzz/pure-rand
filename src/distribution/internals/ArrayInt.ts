/**
 * An ArrayInt represents an integer larger than what can be represented in classical JavaScript.
 * The values stored in data must be in the range [0, 0xffffffff].
 *
 * @example
 * ```js
 * { sign:  1, data: [ 42 ] } // = 42
 * { sign: -1, data: [ 42 ] } // = -42
 * { sign: -1, data: [ 5, 42 ] } // = -1 * (5 * 2**32 + 42)
 * { sign: -1, data: [ 1, 5, 42 ] } // = -1 * (1 * 2**64 + 5 * 2**32 + 42)
 * ```
 */
export type ArrayInt = {
  /**
   * Sign of the represented number
   */
  sign: -1 | 1;
  /**
   * Value of the number, must only contain numbers in the range [0, 0xffffffff]
   */
  data: number[];
};

/**
 * Add two ArrayInt
 * @internal
 */
export function addArrayIntToNew(arrayIntA: ArrayInt, arrayIntB: ArrayInt): ArrayInt {
  if (arrayIntA.sign !== arrayIntB.sign) {
    return substractArrayIntToNew(arrayIntA, { sign: -arrayIntB.sign as -1 | 1, data: arrayIntB.data });
  }
  const data: number[] = [];
  let reminder = 0;
  const dataA = arrayIntA.data;
  const dataB = arrayIntB.data;
  for (let indexA = dataA.length - 1, indexB = dataB.length - 1; indexA >= 0 || indexB >= 0; --indexA, --indexB) {
    const vA = indexA >= 0 ? dataA[indexA] : 0;
    const vB = indexB >= 0 ? dataB[indexB] : 0;
    const current = vA + vB + reminder;
    data.push(current >>> 0);
    reminder = ~~(current / 0x100000000);
  }
  if (reminder !== 0) {
    data.push(reminder);
  }
  return { sign: arrayIntA.sign, data: data.reverse() };
}

/**
 * Add one to a given positive ArrayInt
 * @internal
 */
export function addOneToPositiveArrayInt(arrayInt: ArrayInt): ArrayInt {
  arrayInt.sign = 1; // handling case { sign: -1, data: [0,...,0] }
  const data = arrayInt.data;
  for (let index = data.length - 1; index >= 0; --index) {
    if (data[index] === 0xffffffff) {
      data[index] = 0;
    } else {
      data[index] += 1;
      return arrayInt;
    }
  }
  data.unshift(1);
  return arrayInt;
}

/** @internal */
function isStrictlySmaller(dataA: number[], dataB: number[]): boolean {
  const maxLength = Math.max(dataA.length, dataB.length);
  for (let index = 0; index < maxLength; ++index) {
    const indexA = index + dataA.length - maxLength;
    const indexB = index + dataB.length - maxLength;
    const vA = indexA >= 0 ? dataA[indexA] : 0;
    const vB = indexB >= 0 ? dataB[indexB] : 0;
    if (vA < vB) return true;
    if (vA > vB) return false;
  }
  return false;
}

/**
 * Substract two ArrayInt
 * @internal
 */
export function substractArrayIntToNew(arrayIntA: ArrayInt, arrayIntB: ArrayInt): ArrayInt {
  if (arrayIntA.sign !== arrayIntB.sign) {
    return addArrayIntToNew(arrayIntA, { sign: -arrayIntB.sign as -1 | 1, data: arrayIntB.data });
  }
  const dataA = arrayIntA.data;
  const dataB = arrayIntB.data;
  if (isStrictlySmaller(dataA, dataB)) {
    const out = substractArrayIntToNew(arrayIntB, arrayIntA);
    out.sign = -out.sign as -1 | 1;
    return out;
  }
  const data: number[] = [];
  let reminder = 0;
  for (let indexA = dataA.length - 1, indexB = dataB.length - 1; indexA >= 0 || indexB >= 0; --indexA, --indexB) {
    const vA = indexA >= 0 ? dataA[indexA] : 0;
    const vB = indexB >= 0 ? dataB[indexB] : 0;
    const current = vA - vB - reminder;
    data.push(current >>> 0);
    reminder = current < 0 ? 1 : 0;
  }
  return { sign: arrayIntA.sign, data: data.reverse() };
}

/**
 * Trim uneeded zeros in ArrayInt
 * and uniform notation for zero: {sign: 1, data: [0]}
 */
export function trimArrayIntInplace(arrayInt: ArrayInt) {
  const data = arrayInt.data;
  let firstNonZero = 0;
  for (; firstNonZero !== data.length && data[firstNonZero] === 0; ++firstNonZero) {}
  if (firstNonZero === data.length) {
    // only zeros
    arrayInt.sign = 1;
    arrayInt.data = [0];
    return arrayInt;
  }
  data.splice(0, firstNonZero);
  return arrayInt;
}

// Helpers specific to 64 bits versions

/** @internal */
export type ArrayInt64 = Required<ArrayInt> & { data: [number, number] };

/**
 * We only accept safe integers here
 * @internal
 */
export function fromNumberToArrayInt64(out: ArrayInt64, n: number): ArrayInt64 {
  if (n < 0) {
    const posN = -n;
    out.sign = -1;
    out.data[0] = ~~(posN / 0x100000000);
    out.data[1] = posN >>> 0;
  } else {
    out.sign = 1;
    out.data[0] = ~~(n / 0x100000000);
    out.data[1] = n >>> 0;
  }
  return out;
}

/**
 * Substract two ArrayInt of 64 bits on 64 bits.
 * With arrayIntA - arrayIntB >= 0
 * @internal
 */
export function substractArrayInt64(out: ArrayInt64, arrayIntA: ArrayInt64, arrayIntB: ArrayInt64): ArrayInt64 {
  const lowA = arrayIntA.data[1];
  const highA = arrayIntA.data[0];
  const signA = arrayIntA.sign;
  const lowB = arrayIntB.data[1];
  const highB = arrayIntB.data[0];
  const signB = arrayIntB.sign;

  // Requirement: arrayIntA - arrayIntB >= 0
  out.sign = 1;

  if (signA === 1 && signB === -1) {
    // Operation is a simple sum of arrayIntA + abs(arrayIntB)
    const low = lowA + lowB;
    const high = highA + highB + (low > 0xffffffff ? 1 : 0);
    out.data[0] = high >>> 0;
    out.data[1] = low >>> 0;
    return out;
  }
  // signA === -1 with signB === 1 is impossible given: arrayIntA - arrayIntB >= 0

  // Operation is a substraction
  let lowFirst = lowA;
  let highFirst = highA;
  let lowSecond = lowB;
  let highSecond = highB;
  if (signA === -1) {
    lowFirst = lowB;
    highFirst = highB;
    lowSecond = lowA;
    highSecond = highA;
  }
  let reminderLow = 0;
  let low = lowFirst - lowSecond;
  if (low < 0) {
    reminderLow = 1;
    low = low >>> 0;
  }
  out.data[0] = highFirst - highSecond - reminderLow;
  out.data[1] = low;
  return out;
}
