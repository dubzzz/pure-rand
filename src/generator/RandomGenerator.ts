export interface RandomGenerator {
  /** Produce a fully independent clone of the current instance */
  clone(): RandomGenerator;
  /**
   * Generate next random value along with the next generator (does not impact current instance).
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  next(): [number, RandomGenerator];
  /** Compute the jumped generator (does not impact current instance) */
  jump?(): RandomGenerator;
  /**
   * Generate next value BUT alters current generator.
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  unsafeNext(): number;
  /** Jump current generator */
  unsafeJump?(): void;
}

export function unsafeGenerateN(rng: RandomGenerator, num: number): number[] {
  const out: number[] = [];
  for (let idx = 0; idx != num; ++idx) {
    out.push(rng.unsafeNext());
  }
  return out;
}

export function generateN(rng: RandomGenerator, num: number): [number[], RandomGenerator] {
  const nextRng = rng.clone();
  const out = unsafeGenerateN(nextRng, num);
  return [out, nextRng];
}

export function unsafeSkipN(rng: RandomGenerator, num: number): void {
  for (let idx = 0; idx != num; ++idx) {
    rng.unsafeNext();
  }
}

export function skipN(rng: RandomGenerator, num: number): RandomGenerator {
  const nextRng = rng.clone();
  unsafeSkipN(nextRng, num);
  return nextRng;
}
