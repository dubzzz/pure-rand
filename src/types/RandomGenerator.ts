export interface RandomGenerator {
  /** Produce a fully independent clone of the current instance */
  clone(): RandomGenerator;
  /**
   * Generate next random value along with the next generator (does not impact current instance).
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  next(): [number, RandomGenerator];
  /** Compute the jumped generator (does not impact current instance) */
  jump?(): RandomGenerator;
  /**
   * Generate next value BUT alters current generator.
   * Values uniform in range -0x8000_0000 (included) to 0x7fff_ffff (included)
   */
  unsafeNext(): number;
  /** Jump current generator */
  unsafeJump?(): void;
  /** Access to the internal state of a RandomGenerator in a read-only fashion */
  getState?(): readonly number[];
}
