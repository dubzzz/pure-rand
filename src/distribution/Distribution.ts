import { RandomGenerator } from '../generator/RandomGenerator';

/**
 * Generate random value based on a given RandomGenerator.
 * Return the generated value and an offsetted version of the RandomGenerator (never alters the source rng).
 * @public
 */
export type Distribution<T> = (rng: RandomGenerator) => [T, RandomGenerator];
