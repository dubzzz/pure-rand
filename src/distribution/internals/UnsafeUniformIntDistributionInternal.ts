import { RandomGenerator } from '../../generator/RandomGenerator';

/**
 * Uniformly generate number in range [0 ; rangeSize[
 * With rangeSize <= 0x100000000
 * @internal
 */
export function unsafeUniformIntDistributionInternal(rangeSize: number, rng: RandomGenerator): number {
  // Range provided by the RandomGenerator is large enough,
  // given rangeSize <= 0x100000000 and RandomGenerator is uniform on 0x100000000 values
  const MaxAllowed = rangeSize > 2 ? ~~(0x100000000 / rangeSize) * rangeSize : 0x100000000;
  let deltaV = rng.unsafeNext() + 0x80000000;
  while (deltaV >= MaxAllowed) {
    deltaV = rng.unsafeNext() + 0x80000000;
  }
  return deltaV % rangeSize;
}
