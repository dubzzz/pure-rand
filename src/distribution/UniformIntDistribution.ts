import { Distribution } from './Distribution';
import { RandomGenerator } from '../generator/RandomGenerator';
import { unsafeUniformIntDistribution } from './UnsafeUniformIntDistribution';

/**
 * Uniformly generate random integer values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 *
 * @public
 */
function uniformIntDistribution(from: number, to: number): Distribution<number>;
/**
 * Uniformly generate random integer values between `from` (included) and `to` (included)
 *
 * @param from - Lower bound of the range (included)
 * @param to - Upper bound of the range (included)
 * @param rng - Instance of RandomGenerator to extract random values from
 *
 * @public
 */
function uniformIntDistribution(from: number, to: number, rng: RandomGenerator): [number, RandomGenerator];
function uniformIntDistribution(from: number, to: number, rng?: RandomGenerator) {
  if (rng != null) {
    const nextRng = rng.clone();
    return [unsafeUniformIntDistribution(from, to, nextRng), nextRng];
  }
  return function (rng: RandomGenerator) {
    const nextRng = rng.clone();
    return [unsafeUniformIntDistribution(from, to, nextRng), nextRng];
  };
}

export { uniformIntDistribution };
