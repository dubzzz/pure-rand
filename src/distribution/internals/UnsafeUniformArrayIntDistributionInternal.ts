import { RandomGenerator } from '../../generator/RandomGenerator';
import { ArrayInt } from './ArrayInt';
import { unsafeUniformIntDistributionInternal } from './UnsafeUniformIntDistributionInternal';

/**
 * Uniformly generate ArrayInt in range [0 ; rangeSize[
 *
 * @remarks
 * In the worst case scenario it may discard half of the randomly generated value.
 * Worst case being: most significant number is 1 and remaining part evaluates to 0.
 *
 * @internal
 */
export function unsafeUniformArrayIntDistributionInternal(
  out: ArrayInt['data'],
  rangeSize: ArrayInt['data'],
  rng: RandomGenerator
): ArrayInt['data'] {
  const rangeLength = rangeSize.length;

  // We iterate until we find a valid value for arrayInt
  while (true) {
    // We compute a new value for arrayInt
    for (let index = 0; index !== rangeLength; ++index) {
      const indexRangeSize = index === 0 ? rangeSize[0] + 1 : 0x100000000;
      const g = unsafeUniformIntDistributionInternal(indexRangeSize, rng);
      out[index] = g;
    }

    // If in the correct range we can return it
    for (let index = 0; index !== rangeLength; ++index) {
      const current = out[index];
      const currentInRange = rangeSize[index];
      if (current < currentInRange) {
        return out; // arrayInt < rangeSize
      } else if (current > currentInRange) {
        break; // arrayInt > rangeSize
      }
    }
    // Otherwise we need to try another one
  }
}
