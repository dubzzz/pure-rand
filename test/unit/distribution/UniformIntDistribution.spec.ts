import * as fc from 'fast-check';
import { mocked } from '../../__test-helpers__/mocked';

import { uniformIntDistribution } from '../../../src/distribution/UniformIntDistribution';
import { RandomGenerator } from '../../../src/generator/RandomGenerator';

import * as UnsafeUniformIntDistributionInternalMock from '../../../src/distribution/internals/UnsafeUniformIntDistributionInternal';
import * as UnsafeUniformArrayIntDistributionInternalMock from '../../../src/distribution/internals/UnsafeUniformArrayIntDistributionInternal';
jest.mock('../../../src/distribution/internals/UnsafeUniformIntDistributionInternal');
jest.mock('../../../src/distribution/internals/UnsafeUniformArrayIntDistributionInternal');

function buildUniqueRng(clonedRng?: RandomGenerator) {
  if (clonedRng !== undefined) {
    return {
      clone() {
        return clonedRng;
      },
    } as RandomGenerator;
  }
  return {} as RandomGenerator;
}
function clean() {
  jest.resetAllMocks();
  jest.clearAllMocks();
}

beforeEach(clean);
describe('uniformIntDistribution', () => {
  describe('Small ranges (<= 2**32)', () => {
    it('Should call unsafeUniformIntDistributionInternal with correct size of range and source rng', () =>
      fc.assert(
        fc
          .property(settingsArbitrary, (settings) => {
            // Arrange
            const { from, to, rng, clonedRng, unsafeUniformIntDistributionInternal } = mockInternals(settings);

            // Act
            uniformIntDistribution(from, to)(rng);

            // Assert
            expect(unsafeUniformIntDistributionInternal).toHaveBeenCalledTimes(1);
            expect(unsafeUniformIntDistributionInternal).toHaveBeenCalledWith(to - from + 1, clonedRng);
          })
          .beforeEach(clean)
      ));

    it('Should offset by "from" the value produced by unsafeUniformIntDistributionInternal', () =>
      fc.assert(
        fc
          .property(settingsArbitrary, (settings) => {
            // Arrange
            const { from, to, rng, outputs } = mockInternals(settings);

            // Act
            const [v, _nrng] = uniformIntDistribution(from, to)(rng);

            // Assert
            expect(v).toBe(outputs[0] + from);
          })
          .beforeEach(clean)
      ));
  });

  describe('Large ranges (> 2**32)', () => {
    it('Should call unsafeUniformIntDistributionInternal with correct size of range and source rng', () =>
      fc.assert(
        fc
          .property(settingsLargeArbitrary, (settings) => {
            // Arrange
            const { from, to, rng, clonedRng, unsafeUniformArrayIntDistributionInternal } =
              mockLargeInternals(settings);

            // Act
            uniformIntDistribution(from, to)(rng);

            // Assert
            expect(unsafeUniformArrayIntDistributionInternal).toHaveBeenCalledTimes(1);
            expect(unsafeUniformArrayIntDistributionInternal).toHaveBeenCalledWith(
              expect.any(Array),
              expect.any(Array),
              clonedRng
            );
            const params = unsafeUniformArrayIntDistributionInternal.mock.calls[0];
            const rangeSize = params[1];
            expect(rangeSize[0] * 2 ** 32 + rangeSize[1]).toBe(to - from + 1);
          })
          .beforeEach(clean)
      ));

    it('Should offset by "from" the value produced by unsafeUniformIntDistributionInternal', () =>
      fc.assert(
        fc
          .property(settingsLargeArbitrary, (settings) => {
            // Arrange
            const { from, to, rng, outputs } = mockLargeInternals(settings);

            // Act
            const [v, _nrng] = uniformIntDistribution(from, to)(rng);

            // Assert
            expect(v).toBe(outputs[0][0] * 2 ** 32 + outputs[0][1] + from);
          })
          .beforeEach(clean)
      ));
  });

  it('Should return the rng passed to unsafeUniformIntDistributionInternal', () =>
    fc.assert(
      fc
        .property(settingsArbitrary, (settings) => {
          // Arrange
          const { from, to, rng, clonedRng, outputs } = mockInternals(settings);

          // Act
          const [_v1, nrng] = uniformIntDistribution(from, to)(rng);

          // Assert
          expect(nrng).toBe(clonedRng);
        })
        .beforeEach(clean)
    ));

  it('Should be equivalent to call the 2-parameter and 3-parameter', () =>
    fc.assert(
      fc
        .property(settingsArbitrary, (settings) => {
          // Arrange
          const { from, to, rng } = mockInternals(settings);

          // Act
          const [v1, _nrng1] = uniformIntDistribution(from, to)(rng);
          const [v2, _nrng2] = uniformIntDistribution(from, to, rng);

          // Assert
          expect(v1).toBe(v2);
        })
        .beforeEach(clean)
    ));
});

// Helpers

const settingsArbitrary = fc
  .record(
    {
      from: fc.maxSafeInteger(),
      gap: fc.integer({ min: 0, max: 0xffffffff }),
      rangeRandom: fc.nat().noShrink(),
      ctx: fc.context(),
    },
    { withDeletedKeys: false }
  )
  .filter(({ from, gap }) => Number.isSafeInteger(from + gap));

type SettingsType = typeof settingsArbitrary extends fc.Arbitrary<infer U> ? U : never;

function mockInternals(settings: SettingsType) {
  const { unsafeUniformIntDistributionInternal } = mocked(UnsafeUniformIntDistributionInternalMock);

  const { from, gap, rangeRandom, ctx } = settings;
  const to = from + gap;
  const clonedRng = buildUniqueRng();
  const rng = buildUniqueRng(clonedRng);
  const outputs: number[] = [];
  unsafeUniformIntDistributionInternal.mockImplementation((rangeSize) => {
    const out = rangeRandom % rangeSize;
    ctx.log(`unsafeUniformIntDistributionInternal(${rangeSize}) -> ${out}`);
    outputs.push(out);
    return out;
  });

  return { from, to, rng, clonedRng, outputs, unsafeUniformIntDistributionInternal };
}

const settingsLargeArbitrary = fc
  .record(
    {
      from: fc.maxSafeInteger(),
      gap: fc.integer({ min: 0x100000000, max: Number.MAX_SAFE_INTEGER }),
      rangeRandom: fc.nat().noShrink(),
      ctx: fc.context(),
    },
    { withDeletedKeys: false }
  )
  .filter(({ from, gap }) => Number.isSafeInteger(from + gap));

type SettingsLargeType = typeof settingsLargeArbitrary extends fc.Arbitrary<infer U> ? U : never;

function mockLargeInternals(settings: SettingsLargeType) {
  const { unsafeUniformArrayIntDistributionInternal } = mocked(UnsafeUniformArrayIntDistributionInternalMock);

  const { from, gap, rangeRandom, ctx } = settings;
  const to = from + gap;
  const clonedRng = buildUniqueRng();
  const rng = buildUniqueRng(clonedRng);
  const outputs: number[][] = [];
  unsafeUniformArrayIntDistributionInternal.mockImplementation((rangeSize) => {
    const out = rangeSize.map((r) => rangeRandom % (r || 1));
    ctx.log(`unsafeUniformArrayIntDistributionInternal(${JSON.stringify(rangeSize)}) -> ${JSON.stringify(out)}`);
    outputs.push(out);
    return out;
  });

  return { from, to, rng, clonedRng, outputs, unsafeUniformArrayIntDistributionInternal };
}
