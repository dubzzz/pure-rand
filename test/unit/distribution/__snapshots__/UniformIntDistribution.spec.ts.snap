// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uniformIntDistribution Should not change its output in range (-9007199254740991, 9007199254740991) except for major bumps 1`] = `
Array [
  8737460674930689,
  -2631724699627519,
  -4869445551763455,
  8293357854228481,
  5809860774060033,
  -1881172549904383,
  -7398457846580223,
  -5844702921965567,
  5236253499641857,
  2596198112838657,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 6) except for major bumps 1`] = `
Array [
  4,
  3,
  0,
  2,
  1,
  2,
  6,
  3,
  4,
  5,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 7) except for major bumps 1`] = `
Array [
  4,
  7,
  5,
  0,
  3,
  3,
  3,
  7,
  1,
  3,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 8) except for major bumps 1`] = `
Array [
  8,
  0,
  2,
  6,
  7,
  6,
  7,
  1,
  1,
  0,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 4294967294) except for major bumps 1`] = `
Array [
  2357136044,
  2546248239,
  3071714933,
  3626093760,
  2588848963,
  3684848379,
  2340255427,
  3638918503,
  1819583497,
  2678185683,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 4294967295) except for major bumps 1`] = `
Array [
  2357136044,
  2546248239,
  3071714933,
  3626093760,
  2588848963,
  3684848379,
  2340255427,
  3638918503,
  1819583497,
  2678185683,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 4294967296) except for major bumps 1`] = `
Array [
  189112320,
  554379264,
  1095999488,
  1298663424,
  858602496,
  3171780608,
  3693446144,
  708411392,
  1327118336,
  404687872,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 1099511627774) except for major bumps 1`] = `
Array [
  741289830400,
  506149267456,
  291457769472,
  841166684160,
  41339999232,
  91856056320,
  155904065536,
  300906252288,
  379144339456,
  251166092288,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 1099511627775) except for major bumps 1`] = `
Array [
  741280622592,
  506137268224,
  291447656448,
  841157541888,
  41332891648,
  91845220352,
  155896724480,
  300891291648,
  379128172544,
  251159659520,
]
`;

exports[`uniformIntDistribution Should not change its output in range (0, 1099511627776) except for major bumps 1`] = `
Array [
  741271414784,
  506125268992,
  291437543424,
  841148399616,
  41325784064,
  91834384384,
  155889383424,
  300876331008,
  379112005632,
  251153226752,
]
`;

exports[`uniformIntDistribution Should not change its output in range (48, 69) except for major bumps 1`] = `
Array [
  58,
  55,
  49,
  62,
  67,
  51,
  63,
  59,
  65,
  67,
]
`;
